# 1. Постановка задачи

На рынке представлено специализированное технологическое оборудование двух марок:  
$M1$ и $M2$. На производственном предприятии в данный момент используется оборудование  
марки $M1$, возраст которого составляет $2[года]$. Остаточная стоимость оборудования и  
годовая стоимость обслуживания оборудования в зависимости от срока эксплуатации  
приведены в табл. $1$. Стоимость инструктажа персонала производственной линии при  
смене типа оборудования $550[тыс. руб.]$ (в любом случае, независимо от того, был ли ранее  
опыт работы с оборудованием соответствующей марки), стоимость нового оборудования  
марки $M1$ — $13500[тыс. руб.]$, а $M2$ — $9500[тыс. руб.]$  
Необходимо определить оптимальную стратегию замены оборудования на ближайшие $6[лет]$,  
исходя из того, что через $6[лет]$ оборудование будет реализовано по остаточной стоимости.  
Определить границы изменения стоимости нового оборудования марки $M1$, в которых  
найденная стратегия остается оптимальной.

<table style="display: inline-block">
    <caption>Таблица $1$: Затраты, связанные с эксплуатацией оборудования</caption>
    <tr>
        <th>Возраст</th>
        <th colspan=2>$M1$</th>
        <th colspan=2>$M2$</th>
    </tr>
    <tr>
        <td> </td>
        <td>Остаточная стоимость,<br>$[тыс. руб.]$</td>
        <td>Обслуживание,<br>$[тыс. руб.]$</td>
        <td>Остаточная стоимость,<br>$[тыс. руб.]$</td>
        <td>Обслуживание,<br>$[тыс. руб.]$</td>
    </tr>
    <tr>
        <td>$0$</td>
        <td>$-$</td>
        <td>$400$</td>
        <td>$-$</td>
        <td>$600$</td>
    </tr>
    <tr>
        <td>$1$</td>
        <td>$10800$</td>
        <td>$480$</td>
        <td>$8550$</td>
        <td>$720$</td>
    </tr>
    <tr>
        <td>$2$</td>
        <td>$9720$</td>
        <td>$576$</td>
        <td>$7695$</td>
        <td>$936$</td>
    </tr>
    <tr>
        <td>$3$</td>
        <td>$8748$</td>
        <td>$691$</td>
        <td>$5540$</td>
        <td>$1216$</td>
    </tr>
    <tr>
        <td>$4$</td>
        <td>$7000$</td>
        <td>$829$</td>
        <td>$4432$</td>
        <td>$1460$</td>
    </tr>
    <tr>
        <td>$5$</td>
        <td>$5598$</td>
        <td>$995$</td>
        <td>$3545$</td>
        <td>$1752$</td>
    </tr>
    <tr>
        <td>$6$</td>
        <td>$4478$</td>
        <td>$1194$</td>
        <td>$2800$</td>
        <td>$2100$</td>
    </tr>
    <tr>
        <td>$7$</td>
        <td>$3583$</td>
        <td>$1433$</td>
        <td>$2200$</td>
        <td>$2100$</td>
    </tr>
    <tr>
        <td>$8$</td>
        <td>$2866$</td>
        <td>$1500$</td>
        <td>$2200$</td>
        <td>$2100$</td>
    </tr>
    <tr>
        <td>$9$</td>
        <td>$2800$</td>
        <td>$1500$</td>
        <td>$2200$</td>
        <td>$2100$</td>
    </tr>
    <tr>
        <td>$10$</td>
        <td>$2800$</td>
        <td>$1500$</td>
        <td>$2200$</td>
        <td>$2100$</td>
    </tr>
</table>

# 2. Формализация задачи

Опишем процесс решения:

- Этапами в данном случае будут являться года $(i)$;
- Выигрыш - это минимизация затрат на эксплуатацию и замену оборудования;
- Управление - это выбор одной из трёх опций $(u_i\in\{0,1,2\})$:
    - не менять оборудование;
    - заменить оборудование на новое $M1$;
    - заменить оборудование на новое $M2$.
- Состоянием является кортеж из двух параметров оборудования: марки и возраста $(m, e)$.

Затраты на эксплуатацию обозначим через $h^{(j)}_i$, остаточную стоимость через $r^{(j)}_i$,  
затраты на смену через $с^{(j)}$, обучение через $t^{(j)}$, где $j=\{1,2\}$.  
Также введём функцию оплаты за обучение сотрудников в зависимости от того, какое оборудование  
было на предыдущем этапе.  
Помимо прочего обозначим за $F$ (без зависимостей) общую составляющую уравнений.

$g^{(j)}(m_{i-1})=\begin{cases}0, m_{i-1}=j\\t^{j}, \text{otherwise}\end{cases}$

$F:=\begin{cases}
h^{(m_i)}_{e_i}, u = 0\\
c^{(1)}+h^{(1)}_0-r^{(m_{i-1})}_{e_i}+g^{(1)}(m_{i-1}), u = 1\\
c^{(2)}+h^{(2)}_0-r^{(m_{i-1})}_{e_i}+g^{(2)}(m_{i-1}), u = 2
\end{cases}$

$W_i(m_i,e_i)=\underset{u_i\in\{u|u\in\{0,1,2\}\}}{min}\{F
+W_{i+1}(m_{i+1},e_{i+1})\}, i=\overline{1,4}$

$W_i(m_i,e_i)=\underset{u_i\in\{u|u\in\{0,1,2\}\}}{min}\{F+r^{m_5}_{e_{i+1}}\}, i=5$

Задача будет решаться рекурсивно в двух вариациях:
- без использования кэширования;
- с использованием кэширования.

Никаких библиотек использоваться не будет, для реализации уравнения Бэллмана  
для данной задачи достаточно наличие:
- цикла;
- условного оператора;
- для реализации с кэшированием также потребуется массив со словарями.

# 3. Решение задачи

In [1]:
# Текущий год, текущая марка, текущий срок эксплуатации (без кэша)
def N(s, m, e):
  if s > y: return 0
  w = None
  for o in [0, 1, 2]: # 0 - не менять, 1 - на новое 1, 2 - на новое 2
    if    o == 0: # Обслуж. + некст_шаг.стоимость
      v = -(m - 2) * m1[e] + (m - 1) * m2[e] + N(s + 1, m, e + 1)
    elif  o == 1: # Замена(1) + [переобуч.] + обслуж. - ост. ст. + некст_шаг.ст.
      v = c[0] + t * (m - 1) + m1[0] - r[e][m - 1] + N(s + 1, 1, 1)
    elif  o == 2: # Замена(2) + [переобуч.] + обслуж. - ост. ст. + некст_шаг.ст.
      v = c[1] - t * (m - 2) + m2[0] - r[e][m - 1] + N(s + 1, 2, 1)
    if s == y:   v -= r[e + 1][m - 1] # ост. ст. в последний год
    if w is None or v < w: w = v
  return w
# Текущий год, текущая марка, текущий срок эксплуатации (с кэшом)
def O(s, m, e, h):
  if s > y:           return 0, None
  if (m, e) in h[s]:  return h[s][m, e]
  w, u = None, None
  for o in [0, 1, 2]: # 0 - не менять, 1 - на новое 1, 2 - на новое 2
    if    o == 0: # Обслуж. + некст_шаг.стоимость
      v = -(m - 2) * m1[e] + (m - 1) * m2[e] + O(s + 1, m, e + 1, h)[0]
    elif  o == 1: # Замена(1) + [переобуч.] + обслуж. - ост. ст. + некст_шаг.ст.
      v = c[0] + t * (m - 1) + m1[0] - r[e][m - 1] + O(s + 1, 1, 1, h)[0]
    elif  o == 2: # Замена(2) + [переобуч.] + обслуж. - ост. ст. + некст_шаг.ст.
      v = c[1] - t * (m - 2) + m2[0] - r[e][m - 1] + O(s + 1, 2, 1, h)[0]
    if s == y:   v -= r[e + 1][m - 1] # ост. ст. в последний год
    if w is None or v < w: w, u = v, o
  h[s][m, e] = w, u
  return h[s][m, e]
# Восстановление результата (берём из кэша, ничего заново не перерешиваем) ;)
def R(h):
  r     = []
  k, v  = zip(*h[1].items())
  k, v  = k[0], v[0]
  r.append(v[1])
  m, e  = k[0], k[1]
  for i in range(2, len(h)):
    if    v[1] == 0 : k, v = (m, e + 1), h[i][m, e + 1]
    elif  v[1] == 1 : k, v = (1, 1), h[i][1, 1]
    else            : k, v = (2, 1), h[i][2, 1]
    m, e = k[0], k[1]
    r.append(v[1])
  return r


r   =[[0, 0],
      [10800, 8550],  [9720, 7695], [8748, 5540], [7000, 4432], [5598, 3545],
      [4478, 2800], [3583, 2200], [2866, 2200], [2800, 2200], [2800, 2200]]
m1  = [400, 480, 576, 691, 829, 995, 1194, 1433, 1500, 1500, 1500]
m2  = [600, 720, 936, 1216, 1460, 1752, 2100, 2100, 2100, 2100, 2100]
c   = [13500, 9500]
t   = 550
y   = 5
h   = [{} for _ in range(y + 1)]

In [2]:
%timeit O(1,1,2,h)
%timeit N(1,1,2)

285 ns ± 10 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
232 µs ± 2.44 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [3]:
y = 8
%timeit O(1,1,2,h)
%timeit N(1,1,2)

295 ns ± 8 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
6.2 ms ± 36.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [4]:
y = 5
h   = [{} for _ in range(y + 1)]

res = O(1, 1, 2, h)
ctr = R(h)
print(f"Optimal cost(cache):\t{res[0]}[t.r]")
res = N(1, 1, 2)
print(f"Optimal cost(naive):\t{res}[t.r]")
print(f"Control path(cache):\t{ctr}")

Optimal cost(cache):	-1397[t.r]
Optimal cost(naive):	-1397[t.r]
Control path(cache):	[0, 2, 2, 2, 0]


# 4. Результат

Как видно, при оптимальном плане $0,\ 2,\ 2,\ 2,\ 0$, т.е.:
- остаться с оборудованием марки $M1$;
- перейти на новое оборудование марки $M2$;
- перейти на новое оборудование марки $M2$;
- перейти на новое оборудование марки $M2$;
- остаться с оборудованием марки $M2$;

после чего продать оборудование марки $M2$ (никакого выбора на этом году не предполагается),  
можно получить $1397[т.р.]$, т.е. вместо затрат получаем прибыль.

Также видно, что наивная реализация с увеличением количества периодов начинает  
чудовищно проигрывать реализации с мемоизацией (единицы$[ms]$ против сотен$[ns]$).

# 5. Исследование решения

Определить границы изменения стоимости нового оборудования марки $M1$, в которых  
найденная стратегия остается оптимальной.

Из представленного управления очевидно, что верхняя граница стоимости нового  
оборудования марки $M1$, в которых найденная стратегия остается оптимальной,  
является $\infty$. Для определения нижней границы будем уменьшать стоимость  
нового оборудования марки $M1$ на $1[т.р.]$.

In [5]:
c   = [13500, 9500]
h   = [{} for _ in range(y + 1)]
res = O(1, 1, 2, h)
b   = 13500
ctr = [0, 2, 2, 2, 0]

while b >= 0 and res[0] == -1397 and ctr == [0, 2, 2, 2, 0]:
    b   -= 1
    c    = [b, 9500]
    h    = [{} for _ in range(y + 1)]
    res  = O(1, 1, 2, h)
    ctr  = R(h)

print(f'M1 bottom line:\t{b}\nOptimal cost:\t{res[0]}\nControl path:\t{ctr}')

M1 bottom line:	12542
Optimal cost:	-1398
Control path:	[0, 1, 0, 1, 2]


Для того, чтобы стратегия поменялась, равно как и выигрыш, необходимо  
изменить стоимость замены оборудования марки $M1$ на $12542[т.р.]$.